R/gm-auth.R

```diff
@@ -0,0 +1,207 @@
+#' ## otherwise, go to browser for authentication and authorization
+#' gm_auth()
+#'
+#' ## force use of a token associated with a specific email
+#' gm_auth(email = "jim@example.com")
+#'
+#' ## force a menu where you can choose from existing tokens or
+#' ## choose to get a new one
+#' gm_auth(email = NA)
+#'
+#' ## use a 'read only' scope, so it's impossible to change data
+#' gm_auth(
+#'   scopes = "https://www.googleapis.com/auth/gmail.readonly"
+#' )
+#'
+#' ## use a service account token
+#' gm_auth(path = "foofy-83ee9e7c9c48.json")
+#' }
+gm_auth <- function(email = NULL,
+                    path = NULL,
+                    scopes = c(
+                      "https://www.googleapis.com/auth/gmail.readonly",
+                      "https://www.googleapis.com/auth/gmail.modify",
+                      "https://www.googleapis.com/auth/gmail.compose",
+                      "https://mail.google.com/"
+                    ),
+                    cache = gargle::gargle_oauth_cache(),
+                    use_oob = gargle::gargle_oob_default(),
+                    token = NULL) {
+  cred <- gargle::token_fetch(
+    scopes = scopes,
+    app = gm_oauth_app(),
+    email = email,
+    path = path,
+    package = "gmailr",
+    cache = cache,
+    use_oob = use_oob,
+    token = token
+  )
+  if (!inherits(cred, "Token2.0")) {
+    stop(
+      "Can't get Google credentials.\n",
+      "Are you running gmailr in a non-interactive session? Consider:\n",
+      "  * Call `gm_auth()` directly with all necessary specifics.\n",
+      call. = FALSE
+    )
+  }
+  .auth$set_cred(cred)
+  .auth$set_auth_active(TRUE)
+
+  invisible()
+}
+
+#' Clear current token
+#'
+#' Clears any currently stored token. The next time gmailr needs a token, the
+#' token acquisition process starts over, with a fresh call to [gm_auth()] and,
+#' therefore, internally, a call to [gargle::token_fetch()]. Unlike some other
+#' packages that use gargle, gmailr is not usable in a de-authorized state.
+#' Therefore, calling `gm_deauth()` only clears the token, i.e. it does NOT
+#' imply that subsequent requests are made with an API key in lieu of a token.
+#'
+#' @family auth functions
+#' @export
+#' @examples
+#' \dontrun{
+#' gm_deauth()
+#' }
+gm_deauth <- function() {
+  .auth$clear_cred()
+  invisible()
+}
+
+#' Produce configured token
+#'
+#' @eval gargle:::PREFIX_token_description(gargle_lookup_table)
+#' @eval gargle:::PREFIX_token_return()
+#'
+#' @family low-level API functions
+#' @export
+#' @examples
+#' \dontrun{
+#' gm_token()
+#' }
+gm_token <- function() {
+  if (!gm_has_token()) {
+    gm_auth()
+  }
+  httr::config(token = .auth$cred)
+}
+
+#' Is there a token on hand?
+#'
+#' Reports whether gmailr has stored a token, ready for use in downstream
+#' requests.
+#'
+#' @return Logical.
+#' @export
+#'
+#' @examples
+#' gm_has_token()
+gm_has_token <- function() {
+  inherits(.auth$cred, "Token2.0")
+}
+
+#' Edit auth configuration
```

Comment:
I think more templating is available than you are taking advantage of:

https://github.com/r-dbi/bigrquery/blob/5f4322866780badd5aec58e0e16199a7844c1ba5/R/bq-auth.R#L133-L137

That also evolved/improved since the initial gargle PR.